homeassistant/components/sharkiq/__init__.py

```diff
@@ -0,0 +1,119 @@
+"""Shark IQ Integration."""
+
+import asyncio
+
+import async_timeout
+from sharkiqpy import (
+    AylaApi,
+    SharkIqAuthError,
+    SharkIqAuthExpiringError,
+    SharkIqNotAuthedError,
+    get_ayla_api,
+)
+import voluptuous as vol
+
+from homeassistant import exceptions
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+
+from .const import API_TIMEOUT, COMPONENTS, DOMAIN, LOGGER
+from .update_coordinator import SharkIqUpdateCoordinator
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+
+class CannotConnect(exceptions.HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+async def async_setup(hass, config):
+    """Set up the sharkiq environment."""
+    hass.data.setdefault(DOMAIN, {})
+    if DOMAIN not in config:
+        return True
+
+
+async def async_connect_or_timeout(ayla_api: AylaApi) -> bool:
+    """Connect to vacuum."""
+    try:
+        with async_timeout.timeout(API_TIMEOUT):
+            LOGGER.debug("Initialize connection to Ayla networks API")
+            await ayla_api.async_sign_in()
+    except SharkIqAuthError as exc:
+        LOGGER.error("Authentication error connecting to Shark IQ api", exc_info=exc)
+        return False
+    except asyncio.TimeoutError as exc:
+        LOGGER.error("Timeout expired", exc_info=exc)
+        raise CannotConnect from exc
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Initialize the sharkiq platform via config entry."""
+    ayla_api = get_ayla_api(
+        username=config_entry.data[CONF_USERNAME],
+        password=config_entry.data[CONF_PASSWORD],
+        websession=hass.helpers.aiohttp_client.async_get_clientsession(),
+    )
+
+    try:
+        if not await async_connect_or_timeout(ayla_api):
+            return False
+    except CannotConnect as exc:
+        raise exceptions.ConfigEntryNotReady from exc
+
+    shark_vacs = await ayla_api.async_get_devices(False)
+    device_names = ", ".join([d.name for d in shark_vacs])
+    LOGGER.debug("Found %d Shark IQ device(s): %s", len(device_names), device_names)
+    coordinator = SharkIqUpdateCoordinator(hass, config_entry, ayla_api, shark_vacs)
+
+    await coordinator.async_refresh()
+
+    if not coordinator.last_update_success:
+        raise exceptions.ConfigEntryNotReady
+
+    hass.data[DOMAIN][config_entry.entry_id] = coordinator
+
+    for component in COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    return True
+
+
+async def async_disconnect_or_timeout(coordinator: SharkIqUpdateCoordinator):
+    """Disconnect to vacuum."""
+    LOGGER.debug("Disconnecting from Ayla Api")
+    with async_timeout.timeout(5):
+        try:
+            await coordinator.ayla_api.async_sign_out()
+        except (SharkIqAuthError, SharkIqAuthExpiringError, SharkIqNotAuthedError):
+            pass
+    return True
```

Comment:
Nothing is checking this return value.

Reply:
It's checked on L60: https://github.com/home-assistant/core/pull/38272/files/e4a4773abb0de36631da3a078d67a1df51dcff34#diff-4f8a584cf1bfa8006ca76edb1af351a4R60

Reply:
No that's `async_connect_or_timeout`. This is `async_disconnect_or_timeout`.

Reply:
:+1:  I can already tell this is going to  be a long day, lol 😕